clients/src/test/java/org/apache/kafka/common/record/CompressionCodecsTest.java

```diff
@@ -0,0 +1,217 @@
+import java.nio.ByteBuffer;
+import java.util.Random;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestWatcher;
+import org.junit.runner.Description;
+import org.xerial.snappy.SnappyInputStream;
+import org.xerial.snappy.SnappyOutputStream;
+
+
+public class CompressionCodecsTest {
+    @Rule
+    public PrintRandomSeed printRule = new PrintRandomSeed();
+    private long seed;
+    private Random random;
+
+    @Before
+    public void setup() {
+        seed = System.currentTimeMillis();
+        random = new Random(seed);
+    }
+
+    @Test
+    public void testByteArrayInputStream() throws Exception {
+        testCodec(
+            new InputStreamWrapper() {
+                @Override
+                public InputStream wrap(InputStream underlying) throws Exception {
+                    //noinspection RedundantCast - serves as assertion
+                    return (ByteArrayInputStream) underlying;
+                }
+            },
+            new OutputStreamWrapper() {
+                @Override
+                public OutputStream wrap(OutputStream underlying) throws Exception {
+                    //noinspection RedundantCast - serves as assertion
+                    return (ByteArrayOutputStream) underlying;
+                }
+            });
+    }
+
+    @Test
+    public void testByteBufferInputStream() throws Exception {
+        byte[] blob = generateRandomData();
+        ByteBufferInputStream is = new ByteBufferInputStream(ByteBuffer.wrap(blob));
+        testCodec(blob, is);
+    }
+
+    @Test(expected = AssertionError.class) //vanilla gzip has a bad impl of available()
+    public void testVanillaGzip() throws Exception {
+        testCodec(
+            new InputStreamWrapper() {
+                @Override
+                public InputStream wrap(InputStream underlying) throws Exception {
+                    return new GZIPInputStream(underlying);
+                }
+            },
+            new OutputStreamWrapper() {
+                @Override
+                public OutputStream wrap(OutputStream underlying) throws Exception {
+                    return new GZIPOutputStream(underlying);
+                }
+            });
+    }
+
+    @Test
+    public void testKafkaGzip() throws Exception {
+        testCodec(
+            new InputStreamWrapper() {
+                @Override
+                public InputStream wrap(InputStream underlying) throws Exception {
+                    return new KafkaGZIPInputStream(underlying);
+                }
+            },
+            new OutputStreamWrapper() {
+                @Override
+                public OutputStream wrap(OutputStream underlying) throws Exception {
+                    return new GZIPOutputStream(underlying); //use vanilla
+                }
+            });
+    }
+
+    @Test
```

Comment:
All these tests could also be parameterized - would be a lot less code, but would likely need reflection to look up the constructor.